src/resolver.cpp

```diff
@@ -133,14 +132,58 @@ Ref locationInManagerContextForEntity(const openassetio::hostApi::ManagerPtr &ma
   return assetPath;
 }
 
+/**
+ * Decorator to stop propagation of all exceptions.
+ *
+ * Exceptions occurring within the wrapped callable will be caught and
+ * logged, and a default-constructed value (of the same type as the
+ * callable's return type) returned instead of propagating the
+ * exception.
+ *
+ * The callable must take no arguments (but may return a value).
+ *
+ * This is needed since USD reacts badly if an exception escapes an Ar
+ * plugin (segfault, sigabrt).
+ *
+ * @tparam Fn Callable type.
+ * @param fn Callable instance.
+ * @param logger OpenAssetIO logger to log exception messages.
+ * @param name Name of function/process that caused the exception, to
+ * append to log messages.
+ * @return Result of callable if no exception occurred, otherwise a
+ * default-constructed object.
+ */
+template <typename Fn>
+auto catchAndLogExceptions(Fn &&fn, const openassetio::log::LoggerInterfacePtr &logger,
+                           const std::string_view name) {
+  try {
+    return fn();
+  } catch (const std::exception &exc) {
+    std::string msg = "OpenAssetIO error in ";
+    msg += name;
+    msg += ": ";
+    msg += exc.what();
+    logger->critical(msg);
+  } catch (...) {
+    std::string msg = "OpenAssetIO error in ";
+    msg += name;
+    msg += ": unknown non-exception type caught";
+    logger->critical(msg);
+  }
+  return std::invoke_result_t<Fn>{};
```

Comment:
[Question] Is simply calling `fn();` off the table? I suspect this is perhaps something to do with handling the `void` case?

Reply:
This is default-constructing an instance of the return type of `fn`, without actually calling `fn` (since that throws an exception if we've got here).

Reply:
Ah, duh!